// lingo3d/lib/interface/ICameraBase.js
import { camNear, camFar } from "../engine/constants";
import { MIN_POLAR_ANGLE, MAX_POLAR_ANGLE } from "../globals";
import { bokehDefault } from "../states/useBokeh";
import { bokehApertureDefault } from "../states/useBokehAperture";
import { bokehFocusDefault } from "../states/useBokehFocus";
import { bokehMaxBlurDefault } from "../states/useBokehMaxBlur";
import { objectManagerDefaults, objectManagerSchema } from "./IObjectManager";
export const cameraBaseSchema = {
    ...objectManagerSchema,
    mouseControl: [Boolean, String],
    fov: Number,
    zoom: Number,
    near: Number,
    far: Number,
    active: Boolean,
    transition: [Boolean, Number],
    bokeh: Boolean,
    bokehFocus: Number,
    bokehMaxBlur: Number,
    bokehAperture: Number,
    minPolarAngle: Number,
    maxPolarAngle: Number,
    minAzimuthAngle: Number,
    maxAzimuthAngle: Number,
    enableDamping: Boolean
};
export const cameraBaseDefaults = {
    ...objectManagerDefaults,
    mouseControl: false,
    fov: 75,
    zoom: 1,
    near: camNear,
    far: camFar,
    active: false,
    transition: [undefined, false],
    bokeh: bokehDefault,
    bokehFocus: bokehFocusDefault,
    bokehMaxBlur: bokehMaxBlurDefault,
    bokehAperture: bokehApertureDefault,
    minPolarAngle: MIN_POLAR_ANGLE,
    maxPolarAngle: MAX_POLAR_ANGLE,
    minAzimuthAngle: [-Infinity, -999999999],
    maxAzimuthAngle: [Infinity, 999999999],
    enableDamping: false
};
