// lingo3d/lib/interface/IPhysics.js
import cubeShape from "../display/core/mixins/PhysicsMixin/cannon/shapes/cubeShape";
import { hideSchema } from "./utils/nonEditorSchemaSet";
export const physicsSchema = {
    maxAngularVelocityX: Number,
    maxAngularVelocityY: Number,
    maxAngularVelocityZ: Number,
    maxVelocityX: Number,
    maxVelocityY: Number,
    maxVelocityZ: Number,
    velocity: Object,
    noTumble: Boolean,
    slippery: Boolean,
    mass: Number,
    physicsGroup: Number,
    ignorePhysicsGroups: Array,
    physics: [String, Boolean],
    physicsShape: Function
};
hideSchema([
    "maxAngularVelocityX",
    "maxAngularVelocityY",
    "maxAngularVelocityZ",
    "maxVelocityX",
    "maxVelocityY",
    "maxVelocityZ",
    "velocity",
    "noTumble",
    "slippery",
    "mass",
    "physicsGroup",
    "ignorePhysicsGroups",
    "physicsShape"
]);
export const physicsDefaults = {
    maxAngularVelocityX: Infinity,
    maxAngularVelocityY: Infinity,
    maxAngularVelocityZ: Infinity,
    maxVelocityX: Infinity,
    maxVelocityY: Infinity,
    maxVelocityZ: Infinity,
    velocity: { x: 0, y: 0, z: 0 },
    noTumble: [undefined, false],
    slippery: [undefined, false],
    mass: [undefined, 1],
    physicsGroup: undefined,
    ignorePhysicsGroups: undefined,
    physics: false,
    physicsShape: cubeShape
};
