// lingo3d/lib/interface/IStaticObjectManaget.js
import { animationMixinDefaults, animationMixinSchema } from "./IAnimationMixin";
import { eventLoopDefaults, eventLoopSchema } from "./IEventLoop";
export const staticObjectManagerSchema = {
    ...eventLoopSchema,
    ...animationMixinSchema,
    onClick: Function,
    onMouseDown: Function,
    onMouseUp: Function,
    onMouseOver: Function,
    onMouseOut: Function,
    onMouseMove: Function,
    onLookToEnd: Function,
    name: String,
    id: String,
    bloom: Boolean,
    reflection: Boolean,
    outline: Boolean,
    visible: Boolean,
    frustumCulled: Boolean,
    metalnessFactor: Number,
    roughnessFactor: Number,
    opacityFactor: Number,
    emissiveIntensityFactor: Number,
    emissiveColorFactor: String,
    colorFactor: String,
    toon: Boolean,
    pbr: Boolean
};
export const staticObjectManagerDefaults = {
    ...eventLoopDefaults,
    ...animationMixinDefaults,
    onClick: undefined,
    onMouseDown: undefined,
    onMouseUp: undefined,
    onMouseOver: undefined,
    onMouseOut: undefined,
    onMouseMove: undefined,
    onLookToEnd: undefined,
    name: "",
    id: undefined,
    bloom: false,
    reflection: false,
    outline: false,
    visible: true,
    frustumCulled: true,
    metalnessFactor: [undefined, 0],
    roughnessFactor: [undefined, 1],
    opacityFactor: [undefined, 1],
    emissiveIntensityFactor: [undefined, 1],
    emissiveColorFactor: [undefined, "#000000"],
    colorFactor: [undefined, "#ffffff"],
    toon: false,
    pbr: false
};
