// lingo3d/lib/interface/ITexturedStandard.js
export const texturedStandardSchema = {
    color: String,
    wireframe: Boolean,
    envMap: String,
    aoMap: String,
    aoMapIntensity: Number,
    bumpMap: String,
    bumpScale: Number,
    displacementMap: String,
    displacementScale: Number,
    displacementBias: Number,
    emissiveColor: String,
    emissiveMap: String,
    emissiveIntensity: Number,
    lightMap: String,
    lightMapIntensity: Number,
    metalnessMap: String,
    metalness: Number,
    roughnessMap: String,
    roughness: Number,
    normalMap: String,
    normalScale: [Object, Number],
    normalMapType: String
};
export const texturedStandardDefaults = {
    color: "#ffffff",
    wireframe: false,
    envMap: undefined,
    aoMap: undefined,
    aoMapIntensity: 1,
    bumpMap: undefined,
    bumpScale: 1,
    displacementMap: undefined,
    displacementScale: 1,
    displacementBias: 0,
    emissiveColor: "#000000",
    emissiveMap: undefined,
    emissiveIntensity: 1,
    lightMap: undefined,
    lightMapIntensity: 1,
    metalnessMap: undefined,
    metalness: 0,
    roughnessMap: undefined,
    roughness: 1,
    normalMap: undefined,
    normalScale: { x: 1, y: 1 },
    normalMapType: undefined
};
