// @lincode/react-reticle/lib/Reticle5.js
import * as React from "react";
import style from "./style";
const Reticle = ({ color = "#000", size = 50, strokeWidth = 6 }) => {
    return (React.createElement("svg", { viewBox: "0 0 1300 1390", xmlns: "http://www.w3.org/2000/svg", xmlSpace: "preserve", style: {
            ...style,
            fillRule: "evenodd",
            clipRule: "evenodd",
            strokeMiterlimit: 10,
            width: size,
            height: size
        } },
        React.createElement("path", { d: "M1059.54 543.417h71.25v79.098", style: {
                fill: "none",
                stroke: color,
                strokeWidth: 1.66 * strokeWidth,
            }, transform: "matrix(6.6724 0 0 5.8528 -6259.2 -3108.113)" }),
        React.createElement("path", { d: "M1059.54 543.417h71.25v79.098", style: {
                fill: "none",
                stroke: color,
                strokeWidth: 1.66 * strokeWidth,
            }, transform: "matrix(0 6.54827 -5.96373 0 4526.65 -6092.038)" }),
        React.createElement("path", { d: "M271.634 164.535v98.922", style: {
                fill: "none",
            }, transform: "matrix(6.22686 0 0 6.11103 -1032.055 -962.451)" }),
        React.createElement("circle", { cx: 272.181, cy: 266.334, r: 7.08, transform: "matrix(3.74514 0 0 3.67547 -361.373 -284.744)" }),
        React.createElement("path", { d: "M271.634 164.535v98.922M279.824 271.309h98.648M271.634 279.887v98.022M263.267 271.309h-97.525", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .41 * strokeWidth,
                strokeLinecap: "round",
                strokeLinejoin: "round",
                strokeMiterlimit: 1.5,
            }, transform: "matrix(6.22686 0 0 6.11103 -1032.055 -962.451)" }),
        React.createElement("path", { d: "M190.071 263.457v16.43", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .41 * strokeWidth,
                strokeLinejoin: "round",
                strokeMiterlimit: 1.5,
            }, transform: "matrix(6.22686 0 0 6.11103 -1007.714 -962.451)" }),
        React.createElement("path", { d: "M203.882 263.457v16.43", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .41 * strokeWidth,
                strokeLinejoin: "round",
                strokeMiterlimit: 1.5,
            }, transform: "matrix(6.22686 0 0 6.11103 -950.004 -962.451)" }),
        React.createElement("path", { d: "M190.071 263.457v16.43", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .43 * strokeWidth,
                strokeLinejoin: "round",
                strokeMiterlimit: 1.5,
            }, transform: "matrix(6.22686 0 0 5.45744 -177.244 -787.108)" }),
        React.createElement("path", { d: "M203.882 263.457v16.43", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .43 * strokeWidth,
                strokeLinejoin: "round",
                strokeMiterlimit: 1.5,
            }, transform: "matrix(6.22686 0 0 5.45744 -117.292 -787.108)" }),
        React.createElement("path", { d: "M263.267 189.719h16.557", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .41 * strokeWidth,
                strokeLinejoin: "round",
                strokeMiterlimit: 1.5,
            }, transform: "matrix(6.22686 0 0 6.11103 -1032.055 -936.302)" }),
        React.createElement("path", { d: "M263.267 189.719h16.557", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .41 * strokeWidth,
                strokeLinejoin: "round",
                strokeMiterlimit: 1.5,
            }, transform: "matrix(6.22686 0 0 6.11103 -1031.507 -791.813)" }),
        React.createElement("path", { d: "M263.267 189.719h16.557", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .41 * strokeWidth,
                strokeLinejoin: "round",
                strokeMiterlimit: 1.5,
            }, transform: "matrix(6.22686 0 0 6.11103 -1032.055 -120.32)" }),
        React.createElement("path", { d: "M263.267 189.719h16.557", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .41 * strokeWidth,
                strokeLinejoin: "round",
                strokeMiterlimit: 1.5,
            }, transform: "matrix(6.22686 0 0 6.11103 -1032.055 20.239)" }),
        React.createElement("path", { d: "M1059.54 543.417h71.25v79.098", style: {
                fill: "none",
                stroke: color,
                strokeWidth: 1.66 * strokeWidth,
            }, transform: "matrix(-6.6724 0 0 -5.8528 7558.57 4493.177)" }),
        React.createElement("path", { d: "M1059.54 543.417h71.25v79.098", style: {
                fill: "none",
                stroke: color,
                strokeWidth: 1.66 * strokeWidth,
            }, transform: "matrix(0 -6.54827 5.96373 0 -3227.272 7477.07)" })));
};
export default Reticle;
