// @lincode/react-reticle/lib/Reticle4.js
import * as React from "react";
import style from "./style";
const Reticle = ({ color = "#000", size = 50, strokeWidth = 6 }) => {
    return (React.createElement("svg", { viewBox: "0 0 1300 1390", xmlns: "http://www.w3.org/2000/svg", xmlSpace: "preserve", style: {
            ...style,
            fillRule: "evenodd",
            clipRule: "evenodd",
            strokeLinecap: "round",
            strokeLinejoin: "round",
            strokeMiterlimit: 1.5,
            width: size,
            height: size
        } },
        React.createElement("path", { d: "M271.634 164.535v98.922", style: {
                fill: "none",
            }, transform: "translate(-1010.529 -958.636) scale(6.09702)" }),
        React.createElement("circle", { cx: 301.524, cy: 315.101, r: 107.813, style: {
                fill: "none",
                stroke: color,
                strokeWidth: .76 * strokeWidth,
            }, transform: "matrix(5.5071 0 0 5.5071 -1014.894 -1039.757)" }),
        React.createElement("circle", { cx: 301.524, cy: 315.101, r: 107.813, style: {
                fill: "none",
                stroke: color,
                strokeWidth: 1.07 * strokeWidth,
            }, transform: "translate(174.055 206.565) scale(1.56217)" }),
        React.createElement("circle", { cx: 272.181, cy: 266.334, r: 7.08, transform: "translate(-300.208 -229.981) scale(3.47502)" }),
        React.createElement("path", { d: "M271.634 164.535v98.922M279.824 271.309h98.648M271.634 279.887v98.022M263.267 271.309h-97.525", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .41 * strokeWidth,
            }, transform: "translate(-1010.529 -958.636) scale(6.09702)" }),
        React.createElement("path", { d: "M190.071 263.457v16.43", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .41 * strokeWidth,
                strokeLinecap: "butt",
            }, transform: "translate(-986.696 -958.636) scale(6.09702)" }),
        React.createElement("path", { d: "M203.882 263.457v16.43", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .41 * strokeWidth,
                strokeLinecap: "butt",
            }, transform: "translate(-930.188 -958.636) scale(6.09702)" }),
        React.createElement("path", { d: "M190.071 263.457v16.43", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .43 * strokeWidth,
                strokeLinecap: "butt",
            }, transform: "matrix(6.09702 0 0 5.44493 -173.542 -783.695)" }),
        React.createElement("path", { d: "M203.882 263.457v16.43", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .43 * strokeWidth,
                strokeLinecap: "butt",
            }, transform: "matrix(6.09702 0 0 5.44493 -114.84 -783.695)" }),
        React.createElement("path", { d: "M263.267 189.719h16.557", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .41 * strokeWidth,
                strokeLinecap: "butt",
            }, transform: "translate(-1010.529 -932.547) scale(6.09702)" }),
        React.createElement("path", { d: "M263.267 189.719h16.557", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .41 * strokeWidth,
                strokeLinecap: "butt",
            }, transform: "translate(-1009.992 -788.395) scale(6.09702)" }),
        React.createElement("path", { d: "M263.267 189.719h16.557", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .41 * strokeWidth,
                strokeLinecap: "butt",
            }, transform: "translate(-1010.529 -118.443) scale(6.09702)" }),
        React.createElement("path", { d: "M263.267 189.719h16.557", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .41 * strokeWidth,
                strokeLinecap: "butt",
            }, transform: "translate(-1010.529 21.801) scale(6.09702)" })));
};
export default Reticle;
