// @lincode/react-reticle/lib/Reticle6.js
import * as React from "react";
import style from "./style";
const Reticle = ({ color = "#000", size = 50, strokeWidth = 6 }) => {
    return (React.createElement("svg", { viewBox: "0 0 1300 1390", xmlns: "http://www.w3.org/2000/svg", xmlSpace: "preserve", style: {
            ...style,
            fillRule: "evenodd",
            clipRule: "evenodd",
            strokeMiterlimit: 10,
            width: size,
            height: size
        } },
        React.createElement("path", { d: "m634.609 1013.1 30.344 30.35", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .37 * strokeWidth,
            }, transform: "translate(-3393.61 -5696.95) scale(6.21622)" }),
        React.createElement("path", { d: "m631.153 922.63 18.693 18.692 18.692-18.692", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .59 * strokeWidth,
            }, transform: "matrix(6.59616 0 0 6.87856 -3639.535 -6308.31)" }),
        React.createElement("path", { d: "m631.153 922.63 18.693 18.692 18.692-18.692", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .62 * strokeWidth,
            }, transform: "matrix(0 6.32103 -6.40146 0 7206.164 -3412.968)" }),
        React.createElement("path", { d: "m631.153 922.63 18.693 18.692 18.692-18.692", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .62 * strokeWidth,
            }, transform: "matrix(-6.32103 0 0 -6.40146 4754.611 7258.088)" }),
        React.createElement("path", { d: "m631.153 922.63 18.693 18.692 18.692-18.692", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .62 * strokeWidth,
            }, transform: "matrix(0 -6.32103 6.40146 0 -5906.187 4802.432)" }),
        React.createElement("path", { d: "M1059.54 543.417h71.25v79.098", style: {
                fill: "none",
                stroke: color,
                strokeWidth: 1.89 * strokeWidth,
            }, transform: "matrix(5.8076 0 0 5.19079 -5366.508 -2674.78)" }),
        React.createElement("path", { d: "M1059.54 543.417h71.25v79.098", style: {
                fill: "none",
                stroke: color,
                strokeWidth: 1.89 * strokeWidth,
            }, transform: "matrix(0 5.8076 -5.19079 0 4021.408 -5321.194)" }),
        React.createElement("path", { d: "M1059.54 543.417h71.25v79.098", style: {
                fill: "none",
                stroke: color,
                strokeWidth: 1.89 * strokeWidth,
            }, transform: "matrix(-5.8076 0 0 -5.19079 6660.38 4066.742)" }),
        React.createElement("path", { d: "M1059.54 543.417h71.25v79.098", style: {
                fill: "none",
                stroke: color,
                strokeWidth: 1.89 * strokeWidth,
            }, transform: "matrix(0 -5.8076 5.19079 0 -2727.542 6713.112)" }),
        React.createElement("path", { d: "m634.609 1043.45 30.227-30.23", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .37 * strokeWidth,
            }, transform: "translate(-3393.61 -5696.95) scale(6.21622)" })));
};
export default Reticle;
