// @lincode/react-reticle/lib/Reticle1.js
import * as React from "react";
import style from "./style";
const Reticle = ({ color = "#000", size = 50, strokeWidth = 3 }) => {
    return (React.createElement("svg", { viewBox: "0 0 1300 1390", xmlns: "http://www.w3.org/2000/svg", xmlSpace: "preserve", style: {
            ...style,
            fillRule: "evenodd",
            clipRule: "evenodd",
            strokeLinecap: "round",
            strokeLinejoin: "round",
            strokeMiterlimit: 1.5,
            width: size,
            height: size
        } },
        React.createElement("path", { d: "M271.634 164.535v98.922", style: {
                fill: "none",
            }, transform: "translate(-1012.855 -962.446) scale(6.11103)" }),
        React.createElement("circle", { cx: 301.524, cy: 315.101, r: 107.813, style: {
                fill: "none",
                stroke: color,
                strokeWidth: .75 * strokeWidth,
            }, transform: "matrix(5.51976 0 0 5.51976 -1017.23 -1043.75)" }),
        React.createElement("circle", { cx: 301.524, cy: 315.101, r: 107.813, style: {
                fill: "none",
                stroke: color,
                strokeWidth: 1.56 * strokeWidth,
            }, transform: "matrix(2.66727 0 0 2.66727 -157.14 -144.93)" }),
        React.createElement("path", { d: "M271.634 164.535v98.922", style: {
                fill: "none",
                stroke: color,
                strokeWidth: 2.93 * strokeWidth,
                strokeLinecap: "square",
            }, transform: "matrix(6.11103 0 0 5.2283 -1012.855 -817.205)" }),
        React.createElement("path", { d: "M279.824 271.309h98.648", style: {
                fill: "none",
                stroke: color,
                strokeWidth: 2.92 * strokeWidth,
                strokeLinecap: "square",
            }, transform: "matrix(5.27213 0 0 6.11103 -695.356 -962.446)" }),
        React.createElement("path", { d: "M271.634 279.887v98.022", style: {
                fill: "none",
                stroke: color,
                strokeWidth: 2.9 * strokeWidth,
                strokeLinecap: "square",
            }, transform: "matrix(6.11103 0 0 5.37705 -1012.855 -685.07)" }),
        React.createElement("path", { d: "M263.267 271.309h-97.525", style: {
                fill: "none",
                stroke: color,
                strokeWidth: 2.91 * strokeWidth,
                strokeLinecap: "square",
            }, transform: "matrix(5.33423 0 0 6.11103 -884.107 -962.446)" })));
};
export default Reticle;
