// @lincode/react-reticle/lib/Reticle2.js
import * as React from "react";
import style from "./style";
const Reticle = ({ color = "#000", size = 50, strokeWidth = 6 }) => {
    return (React.createElement("svg", { viewBox: "0 0 1300 1390", xmlns: "http://www.w3.org/2000/svg", xmlSpace: "preserve", style: {
            ...style,
            fillRule: "evenodd",
            clipRule: "evenodd",
            strokeLinecap: "round",
            strokeLinejoin: "round",
            strokeMiterlimit: 1.5,
            width: size,
            height: size
        } },
        React.createElement("path", { d: "M271.634 164.535v98.922", style: {
                fill: "none",
            }, transform: "translate(-1012.86 -962.446) scale(6.11103)" }),
        React.createElement("circle", { cx: 301.524, cy: 315.101, r: 107.813, style: {
                fill: "none",
                stroke: color,
                strokeWidth: .75 * strokeWidth,
            }, transform: "matrix(5.51976 0 0 5.51976 -1017.235 -1043.75)" }),
        React.createElement("circle", { cx: 272.181, cy: 266.334, r: 7.08, transform: "translate(-1313.15 -1222.62) scale(7.20203)" }),
        React.createElement("path", { d: "M271.634 164.535v98.922", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .64 * strokeWidth,
                strokeLinecap: "butt",
            }, transform: "matrix(6.11103 0 0 4.92166 -1012.86 -766.754)" }),
        React.createElement("path", { d: "M279.824 271.309h98.648", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .64 * strokeWidth,
                strokeLinecap: "butt",
            }, transform: "matrix(4.872 0 0 6.11103 -543.923 -962.446)" }),
        React.createElement("path", { d: "M271.634 279.887v98.022", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .64 * strokeWidth,
                strokeLinecap: "butt",
            }, transform: "matrix(6.11103 0 0 4.8841 -1012.86 -498.778)" }),
        React.createElement("path", { d: "M263.267 271.309h-97.525", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .64 * strokeWidth,
                strokeLinecap: "butt",
            }, transform: "matrix(4.84735 0 0 6.11103 -803.416 -962.446)" })));
};
export default Reticle;
