// @lincode/react-reticle/lib/Reticle7.js
import * as React from "react";
import style from "./style";
const Reticle = ({ color = "#000", size = 50, strokeWidth = 4 }) => {
    return (React.createElement("svg", { viewBox: "0 0 1300 1390", xmlns: "http://www.w3.org/2000/svg", xmlSpace: "preserve", style: {
            ...style,
            fillRule: "evenodd",
            clipRule: "evenodd",
            strokeMiterlimit: 10,
            width: size,
            height: size
        } },
        React.createElement("path", { d: "M1028.53 984.857v86.133", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .43 * strokeWidth,
            }, transform: "translate(-5845.93 -5797.61) scale(6.31162)" }),
        React.createElement("path", { d: "M976.223 919.262h-16.82c-26.597 0-48.191 22.109-48.191 49.341v16.916M911.212 1074.06v18c0 27.23 21.594 49.34 48.191 49.34h16.877M1062.58 1141.4h19.18c26.6 0 48.19-22.11 48.19-49.34v-17.39M1129.95 985.29v-16.687c0-27.232-21.59-49.341-48.19-49.341h-18.95", style: {
                fill: "none",
                stroke: color,
                strokeWidth: 2.89 * strokeWidth,
            }, transform: "matrix(5.82698 0 0 5.69104 -5296.906 -5168.664)" }),
        React.createElement("path", { d: "M986.332 1027.92h85.128", style: {
                fill: "none",
                stroke: color,
                strokeWidth: .43 * strokeWidth,
            }, transform: "translate(-5845.93 -5797.61) scale(6.31162)" })));
};
export default Reticle;
