// @lincode/filetypes/lib/index.js
import { splitFileName } from "@lincode/utils";
export const extensionMap = new Map();
export const getExtensionType = (val) => { var _a, _b; return extensionMap.get((_b = (_a = splitFileName(val)[1]) === null || _a === void 0 ? void 0 : _a.toLowerCase()) !== null && _b !== void 0 ? _b : ""); };
const register = (type, extensions) => {
    for (const ext of extensions)
        extensionMap.set(ext, type);
    return extensions;
};
export const imageExtensions = register("image", [
    "jpg", "jpeg", "gif", "png", "svg", "tiff", "webp", "hdr"
]);
export const audioExtensions = register("audio", [
    "m4a", "mp3", "wav", "wma", "ogg", "flax", "flac", "aac"
]);
export const plainTextExtensions = register("plainText", [
    "ls", "js", "ts", "jsx", "tsx", "py", "txt", "md", "json", "yaml", "c", "cpp", "h", "hpp", "css", "scss", "sass", "html"
]);
export const modelExtensions = register("model", [
    "gltf", "glb", "fbx"
]);
export const sceneExtensions = register("scene", [
    "l3d"
]);
