// @lincode/react-reticle/src/Reticle3.tsx
import * as React from "react"
import { ReticleProps } from "./ReticleProps"
import style from "./style"

const Reticle: React.FC<ReticleProps> = ({ color = "#000", size = 50, strokeWidth = 6 }) => {
    return (
        <svg
            viewBox="0 0 1300 1390"
            xmlns="http://www.w3.org/2000/svg"
            xmlSpace="preserve"
            style={{
                ...style,
                fillRule: "evenodd",
                clipRule: "evenodd",
                strokeLinecap: "round",
                strokeLinejoin: "round",
                strokeMiterlimit: 1.5,
                width: size,
                height: size
            }}
        >
            <path
            d="M271.634 164.535v98.922"
            style={{
                fill: "none",
            }}
            transform="translate(-1000.676 -954.855) scale(6.07684)"
            />
            <circle
            cx={301.524}
            cy={315.101}
            r={107.813}
            style={{
                fill: "none",
                stroke: color,
                strokeWidth: .76 * strokeWidth,
            }}
            transform="matrix(5.48888 0 0 5.48888 -1005.027 -1035.707)"
            />
            <circle
            cx={272.181}
            cy={266.334}
            r={7.08}
            transform="matrix(7.16174 0 0 7.16174 -1299.289 -1213.576)"
            />
            <path d="m649.997 423.846-39.44-378.853h78.88l-39.44 378.853ZM649.998 966.139l39.44 378.853h-78.88l39.44-378.853ZM921.141 685.57l378.853-39.441v78.88l-378.853-39.44ZM378.854 685.57.001 725.012v-78.88l378.853 39.44Z" />
        </svg>
    )
}

export default Reticle
